lib/Sema/CSSimplify.cpp

```diff
@@ -2979,6 +2979,28 @@ bool ConstraintSystem::repairFailures(
     if (lhs->is<InOutType>() || rhs->is<InOutType>())
       break;
 
+    // If this is a failure from a explicit coercion where
+    // both types are BoundedGeneric, try to diagnose the generic mismatch.
+    if (auto last = locator.last()) {
+      if (last->is<LocatorPathElt::ExplicitTypeCoercion>()) {
+        if (lhs->is<BoundGenericType>() && rhs->is<BoundGenericType>()) {
```

Comment:
A better check would be `if (hasConversionOrRestriction(ConversionRestrictionKind::DeepEquality)) break;` because `DeepEquality` restriction is responsible for matching generic parameters and recording appropriate fixes.  You can see the same done here for some other situations e.g. argument-to-parameter mismatches.

Reply:
Humm so if the is a DeepEquallity restriction breaking here will make sure we don't add the ForceDowncast here and defer this to be handled somewhere else or we still would have to make sure that we add the GenericArgumentsMismatch mismatch in the correct place. 
Because if we skip emitting the ForceDowncast fix in match types, if it has a DeepEquallity restriction it probably will defer to the generic mismatch the way it was doing before I've added that fix. So if I'm guessing correctly we wouldn't have to do anything here if we skip adding the ForceDowncast fix in this case right?

Reply:
Yes, that's the idea. `DeepEquality` restriction already does almost exactly what your code does and records `GenericArgumentsMismatch` upon any mismatches it finds. While looking at this I realized that you'd probably want to add a new warning fix to `matchDeepEqualityTypes` to handle cases like `x as S<Int>` where `x` is itself `S<Int>` but that could be addressed separately as a follow-up.